sql: new SHOW SYNTAX statement

knz opened the pull request:
Needed by #19445 - we want to replace the client-side syntax checking by server-side syntax checking.

This PR introduces a server-side statement that checks the syntax of some arbitrary input string.

This PR is the server-side part of that change, I'll have a separate followup PR for the corresponding CLI shell changes.

knz committed:
sql/parser: remove the unused HELP token definition

Release note: none

knz committed:
sql/parser: sort scrub_stmt in the right place

Release note: none

knz committed:
sql/parser: add a missing test

Release note: none

knz committed:
sql: new SHOW SYNTAX statement for smart clients

Release note (sql change): This patch introduces the new statement
SHOW SYNTAX that allows clients to analyze some arbitrary SQL syntax
server-side and either retrieve the (pretty-printed) syntax
decomposition of the string, or the details of the syntax error, if
any.

This is intended e.g. for use in the CockroachDB interactive SQL shell.

knz committed:
sql: make SHOW SYNTAX valid regardless of txn state

This patch promotes SHOW SYNTAX to the same level as
SHOW TRANSACTION STATUS: the statement can be issued
and expected to produce results regardless of the
current txn state -- it's valid even if the txn
is aborted or in state RetryWait. Also it is guaranteed
to not change the txn state.

The patch achieves this by doing some DRY between
the two statements, via a new interface
`parser.ObserverStatement` implemented by both statement
types.

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19793)
<!-- Reviewable:end -->

knz commented:
cc @bdarnell

knz committed:
sql: link to the caller context in pgerror.Error for parse errors

A `pgerror.Error` object is structured and contains a "source
information" field composed of file, line/column number (position),
function.

In an ideal world, this information should be populated as follows:

- for syntax and SQL analysis errors (including planning etc), the
  file/pos/function should be that of the source for the SQL
  text. This means that the fields would be mostly empty for
  interactive clients and single statements, but:

  - the position field should still inform where the error is
    occurring relative to the start of the current SQL text (useful
    for e.g. SQL that spans multiple lines, with comments etc.)
  - if/when the engine supports stored procedures, the function field
    should be populated as well.

- for other errors, the file/pos/function should be that of the
  source code of the engine where the error is raised (e.g. a Go
  file in CockroachDB).

Currently CockroachDB does not do location tracking in the parser so
errors relative to the SQL text cannot refer to the position in the
SQL text. It might be possible to hack this in for "purely" syntax
errors, those generated in the parser itself, but unless/until
locations are stored in the AST/IR it is not possible to equip SQL
errors with the proper location at all.

However, the next best thing is to provide a modicum of context that
informs the user where the parsing is occurring.

For SQL text sent by the client, this information would not be too
interesting, as the context would always be the same: some place in
the Executor. However, there are many "internal" uses of the parser
which also can cause errors. For these, especially during testing, it
is important to determine where the parser causing an error was called
from, as there may be multiple uses of the parser in different places
for a single test / operation.

Prior to this patch, unfortunately the `pgerror.Error` object would be
constructed for SQL syntax errors specifically with the
file/pos/function of just the `parse.Parse()` function, instead of the
call point of that function, which is more interesting.

This patch improves the situation by adding the necessary `depth`
parameters to the various calls so that the `pgerror.Error`
constructor peeks on the call stack one level higher.

Release note (sql change): upon SQL errors, more complete and useful
details are now reported to clients. This is aimed to aid
troubleshooting of CockroachDB.

danhhz commented:
Didn't do a full review, but I like the approach! Also happy that we're addressing this, the client side parsing has tripped me up from time to time.

---

Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 1394 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-KyHhRX4S0G5orr5G8Ag:-KyHhRX4S0G5orr5G8Ah:bphbpjj) ([raw file](https://github.com/cockroachdb/cockroach/blob/db31b3b9d5f0e2c7a9b47d56a907fd7d29cbe553/pkg/sql/executor.go#L1394)):*
> ```Go
> 			return err
> 		}
> 	}
> ```

add a default with a nice error message, please, so it's easy to track this down if someone needs to add another case

---

*[pkg/sql/parser/stmt.go, line 577 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-KyHiA4P6-hE8Bomy9Et:-KyHiA4P6-hE8Bomy9Eu:b-1cfpuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/db31b3b9d5f0e2c7a9b47d56a907fd7d29cbe553/pkg/sql/parser/stmt.go#L577)):*
> ```Go
> 
> // StatementTag returns a short string identifying the type of statement.
> func (*ShowSyntax) StatementTag() string { return "ECHO" }
> ```

is this intentionally still ECHO?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-:-KyHikmySaOxttus2TMC:b-j73zho)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
I defer to someone in SQL land for a full review but this looks good to me. 

Could we also use this interface for completion? For example, `SHOW SYNTAX 'SELECT FROM t'` could return as one of its fields an array of tables whose names start with `t`. Or maybe that should be a different command. 

---

Review status: 0 of 15 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/show_source, line 264 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-KyIHcWssmoALa3Cdsph:-KyIHcWssmoALa3Cdspi:b9wbbd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/db31b3b9d5f0e2c7a9b47d56a907fd7d29cbe553/pkg/sql/logictest/testdata/logic_test/show_source#L264)):*
> ```
> 
> query TT colnames
> SELECT * FROM [SHOW SYNTAX 'select 1; select 2']
> ```

Why the extra `SELECT * FROM` wrapper?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-:-KyIHqswd-bDDMcFs8Se:bwozky)*
<!-- Sent from Reviewable.io -->

knz commented:
The idea to use the API *later* for completion is intruiging and worth exploring. We're miles away from being able to do this though, so it's clearly out of scope for now.

---

Review status: 0 of 15 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/show_source, line 264 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-KyIHcWssmoALa3Cdsph:-KyL7GArWMJRH8UKZs8Y:b-now791) ([raw file](https://github.com/cockroachdb/cockroach/blob/db31b3b9d5f0e2c7a9b47d56a907fd7d29cbe553/pkg/sql/logictest/testdata/logic_test/show_source#L264)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why the extra `SELECT * FROM` wrapper?
</blockquote></details>

This file checks both that the statement is valid and that it behaves sanely as a data source.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-:-KyL75VH6929ts14SsGx:b-lnw27b)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/show.go:
```diff
@@ -659,6 +659,77 @@ func (p *planner) ShowTables(ctx context.Context, n *parser.ShowTables) (planNod
+			query.WriteString(", ")
+		}
+		first = false
+		fmt.Fprintf(&query, "('%s', %s)", field, parser.EscapeSQLString(msg))
```
fmt.Fprintf(&query, "%s('%s', %s)", comma, field, parser.EscapeSQLString(msg))

where comma := "" and then is made ", " after the first iteration

vivekmenezes replied:
hmmm, Actually probably better to add the strings into a []string and then call join after method

vivekmenezes commented on pkg/sql/show.go:
```diff
@@ -659,6 +659,77 @@ func (p *planner) ShowTables(ctx context.Context, n *parser.ShowTables) (planNod
+) error {
+	stmts, err := parser.Parse(stmt)
+	if err != nil {
+		pqErr, _ := pgerror.GetPGCause(err)
```
what's the second return value? is it worth ignoring it?

vivekmenezes reviewed (approved):
Much nicer to have this on the server side! Thanks

knz commented:
Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/executor.go, line 1394 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-KyHhRX4S0G5orr5G8Ag:-L1Sc_xqGSOWqTo5sIwt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db31b3b9d5f0e2c7a9b47d56a907fd7d29cbe553/pkg/sql/executor.go#L1394)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

add a default with a nice error message, please, so it's easy to track this down if someone needs to add another case
</blockquote></details>

Done.

---

*[pkg/sql/show.go, line 675 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-Ky38APBlWvJGlXVy-LR-r6-675:-L1SctVlSy1iIKFu8qPl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db31b3b9d5f0e2c7a9b47d56a907fd7d29cbe553/pkg/sql/show.go#L675)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

hmmm, Actually probably better to add the strings into a []string and then call join after method
</blockquote></details>

Done.

---

*[pkg/sql/show.go, line 689 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-Ky38APBlWvJGlXVy-LR-r6-689:-L1SdwgreGvbXtPzNTdm:b7e94k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/db31b3b9d5f0e2c7a9b47d56a907fd7d29cbe553/pkg/sql/show.go#L689)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

what's the second return value? is it worth ignoring it?
</blockquote></details>

Oversight. Fixed.

---

*[pkg/sql/parser/stmt.go, line 577 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-KyHiA4P6-hE8Bomy9Et:-L1Sdz3JsqpjNP6_Bg9n:b7e94k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/db31b3b9d5f0e2c7a9b47d56a907fd7d29cbe553/pkg/sql/parser/stmt.go#L577)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

is this intentionally still ECHO?
</blockquote></details>

Oversight. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793)*
<!-- Sent from Reviewable.io -->

knz commented:
I have also added the new "release note" strings in each commit message.

knz committed:
sql: simplify the code for SHOW SYNTAX

Using a tip by @vivekmenezes.

Release note: none

knz committed:
sql: add a missing error path for SHOW SYNTAX

To ensure that any further changes to the parser that
can cause non-pgerror errors to crop up don't cause
the statement to panic the node.

Release note: none

knz committed:
sql: avoid an unnecessary string allocation in SHOW SYNTAX

Release note: none

knz commented:
yay it's green

jordanlewis commented on pkg/sql/logictest/testdata/logic_test/show_source:
```diff
@@ -286,3 +286,23 @@ query ITTTTTTTTRTI colnames
 id  type  description  username  status  created  started  finished  modified  fraction_completed  error  coordinator_id
+
+query TT colnames
+SELECT * FROM [SHOW SYNTAX 'select 1; select 2']
```
Could we not just do `SHOW SYNTAX 'select 1; select 2'` here?

jordanlewis commented on pkg/sql/show_syntax.go:
```diff
@@ -0,0 +1,100 @@
+// usually handled as a special case in Executor, but for
+// FROM [SHOW SYNTAX ...] we will arrive here too.
+func (p *planner) ShowSyntax(ctx context.Context, n *tree.ShowSyntax) (planNode, error) {
+	var query bytes.Buffer
```
I think it would help maintainability to put a full example of the query you're constructing in a comment. Otherwise it's hard to figure out the overall flow of what's going on without reading all of the fmt.Fprintf and conditionals below.

jordanlewis commented on pkg/sql/sem/tree/stmt.go:
```diff
@@ -151,6 +151,15 @@ func (l StatementList) Format(buf *bytes.Buffer, f FmtFlags) {
 	}
 }
 
+// ObserverStatement is a pseudo-interface to be implemented by
```
naming nit: pseudo-interface? I would call this a marker interface. Take or leave, especially if we call other things pseudo-interface in the code :)

jordanlewis commented on pkg/sql/show_syntax.go:
```diff
@@ -0,0 +1,100 @@
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package sql
```
Maybe convert this into a pgerror internal error?

jordanlewis reviewed (approved):
Cool! I don't really see the point of the new parse error depth stuff, but LGTM.
<!-- review by re -->

jordanlewis commented on pkg/sql/show_syntax.go:
```diff
@@ -0,0 +1,100 @@
+	if err != nil {
+		pqErr, ok := pgerror.GetPGCause(err)
+		if !ok {
+			return errors.Wrapf(err, "unknown parser error")
```
maybe convert this into a pgerror internal error?

knz commented:
Review status: 0 of 12 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/show_syntax.go, line 33 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-L1SfVYomrwQhhgvQfMC-r15-33:-L1TKeIwsJGua8xsji3R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20627830ca187783c0091126e690bdb6b56267c2/pkg/sql/show_syntax.go#L33)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think it would help maintainability to put a full example of the query you're constructing in a comment. Otherwise it's hard to figure out the overall flow of what's going on without reading all of the fmt.Fprintf and conditionals below.
</blockquote></details>

Done.

---

*[pkg/sql/show_syntax.go, line 57 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-L1SfVYomrwQhhgvQfMC-r15-57:-L1TLFLne1d2HqBYIZKq:b-a9wp08) ([raw file](https://github.com/cockroachdb/cockroach/blob/20627830ca187783c0091126e690bdb6b56267c2/pkg/sql/show_syntax.go#L57)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

maybe convert this into a pgerror internal error?
</blockquote></details>

Solid advice. Done.

---

*[pkg/sql/logictest/testdata/logic_test/show_source, line 291 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-Ky38APA172mEj8AijQM-r15-291:-L1TJxWykDCy3iSfqKtK:b-8f2l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/20627830ca187783c0091126e690bdb6b56267c2/pkg/sql/logictest/testdata/logic_test/show_source#L291)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Could we not just do `SHOW SYNTAX 'select 1; select 2'` here?
</blockquote></details>

Ben was asking the same. This test also checks that the thing is properly recognized as a sub-query, not only as top level statement.

---

*[pkg/sql/sem/tree/stmt.go, line 154 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793#-L1SfVYomrwQhhgvQfMB-r15-154:-L1TKtcQOCECPpUZBTlJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20627830ca187783c0091126e690bdb6b56267c2/pkg/sql/sem/tree/stmt.go#L154)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

naming nit: pseudo-interface? I would call this a marker interface. Take or leave, especially if we call other things pseudo-interface in the code :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19793)*
<!-- Sent from Reviewable.io -->
